web/template.py

```diff
@@ -306,23 +306,38 @@ def paren_expr():
 
         parens = {"(": ")", "[": "]", "{": "}"}
 
-        def get_tokens(text):
+        def get_tokens(text: str):
             """tokenize text using python tokenizer.
             Python tokenizer ignores spaces, but they might be important in some cases.
             This function introduces dummy space tokens when it identifies any ignored space.
             Each token is a storage object containing type, value, begin and end.
             """
-            i = iter([text])
-            readline = lambda: next(i)
-            end = None
-            for t in tokenize.generate_tokens(readline):
-                t = storage(type=t[0], value=t[1], begin=t[2], end=t[3])
-                if end is not None and end != t.begin:
-                    _, x1 = end
-                    _, x2 = t.begin
-                    yield storage(type=-1, value=text[x1:x2], begin=end, end=t.begin)
-                end = t.end
-                yield t
+
+            def tokenize_text(input_text):
+                i = iter([input_text])
+                readline = lambda: next(i)
+                end = None
+                for t in tokenize.generate_tokens(readline):
+                    t = storage(type=t[0], value=t[1], begin=t[2], end=t[3])
+                    if end is not None and end != t.begin:
+                        _, x1 = end
+                        _, x2 = t.begin
+                        yield storage(
+                            type=-1, value=input_text[x1:x2], begin=end, end=t.begin
+                        )
+                    end = t.end
+                    yield t
+
+            try:
+                yield from tokenize_text(text)
+            except tokenize.TokenError as e:
+                # Python 3.12+ compatibility: add a `"` suffix to terminate the string literal.
+                # See https://github.com/webpy/webpy/issues/784.
+                if "unterminated string literal" in str(e):
+                    fixed_text = text + '"'
+                    yield from tokenize_text(fixed_text)
```

Comment:
Does `get_tokens()` raise an exception before returning any tokens? If not, it seems like you're going to get duplicate tokens with the retry strategy.

The current strategy seems like it's going to return a different string of tokens than before. It seems slightly cleaner to  use the same strategy that's currently used for returning white space (token type = -1) and create a token starting at the error position reported in the `TokenError` and running until the end of the string (or perhaps a token of type `token.ERRORTOKEN`).

I think that strategy could probably be used with any type of `TokenError` which is raised without having to check that it's an unterminated string literal.

Note that I haven't tested any of this!

Reply:
I will try this. Thanks!

For what it is worth, at least at the `read_suite()` level, the existing `web.py` code on 3.11, and the PR code running on Python 3.12, produce the same outputs, or seem to, at the `read_suite()` level, though I have not dug into why that 'extra' `"` disappears.

I did take the output from some templates that were crashing on 3.12 and ran them both through 3.11 without the PR code, and 3.12 with the PR code, and ran `sort u` on them in `nvim`, and they were all identical.

For example, on Python 3.12 with this PR's code applied:
```python
Python 3.12.1 (main, Dec 19 2023, 03:09:45) [GCC 12.2.0] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> import web
>>> from web.template import Parser
>>> input_read_section = """$var title: Index
... <p><a href="$back">&larr; Back to Index</a></p>
... <ul>
... $for path, title in docs:
...     <li><a href="$path">$title</a></li>
... </ul>
... """
>>> Parser().read_suite(input_read_section)
[<var: title = join_('Index')>, <line: [t'<p><a href="', $back, t'">&larr; Back to Index</a></p>\n']>, <line: [t'<ul>\n']>, <block: 'for path, title in docs:', [<line: [t'<li><a href="', $path, t'">', $title, t'</a></li>\n']>]>, <line: [t'</ul>\n']>]
```

On Python 3.11, with the existing `web.py` code:
```python
Python 3.11.7 (main, Jan  5 2024, 22:06:16) [GCC 13.2.1 20230801]
Type 'copyright', 'credits' or 'license' for more information
IPython 8.19.0 -- An enhanced Interactive Python. Type '?' for help.

In [1]: import web

In [2]: from web.template import Parser

In [3]: input_read_section = """$var title: Index
   ...: <p><a href="$back">&larr; Back to Index</a></p>
   ...: <ul>
   ...: $for path, title in docs:
   ...:     <li><a href="$path">$title</a></li>
   ...: </ul>
   ...: """

In [4]: Parser().read_suite(input_read_section)
Out[4]: [<var: title = join_('Index')>, <line: [t'<p><a href="', $back, t'">&larr; Back to Index</a></p>\n']>, <line: [t'<ul>\n']>, <block: 'for path, title in docs:', [<line: [t'<li><a href="', $path, t'">', $title, t'</a></li>\n']>]>, <line: [t'</ul>\n']>]
```

I am not in love with adding the single `"`, though. I will look into the suggestion of creating a new token, and put some more time into understanding how these tokens are used, and how much is read (as not every part of a line is sent to this function -- it's the components, as created by `read_node()`, that start with `$` that are sent to `read_expr()`, and hence have `tokenize.generate_tokens()` called on them.

I would have to verify it, but my suspicion is that because the lines seem to end with new line characters, when I append a `"`, it's after a newline that may be getting stripped. But I could be completely wrong. I will both try the suggestion of creating a new token, and also understand why exactly the outputs are identical in 3.11 without the PR, and in 3.12 with the PR.